crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -322,38 +318,45 @@ impl World {
         I: SystemInput + 'static,
         O: 'static,
     {
-        // lookup
+        // Lookup
         let mut entity = self
             .get_entity_mut(id.entity)
             .map_err(|_| RegisteredSystemError::SystemIdNotRegistered(id))?;
 
-        // take ownership of system trait object
+        // Take ownership of system trait object
         let RegisteredSystem {
             mut initialized,
             mut system,
         } = entity
             .take::<RegisteredSystem<I, O>>()
             .ok_or(RegisteredSystemError::Recursive(id))?;
 
-        // run the system
+        // Run the system
         if !initialized {
             system.initialize(self);
             initialized = true;
         }
 
         let result = if system.validate_param(self) {
-            Ok(system.run(input, self))
+            // Wait to run the commands until the system is available again.
+            // This is needed so the systems can recursively run themselves.
+            let ret = system.run_without_applying_deferred(input, self);
+            system.queue_deferred(self.into());
+            Ok(ret)
         } else {
             Err(RegisteredSystemError::InvalidParams(id))
         };
 
-        // return ownership of system trait object (if entity still exists)
+        // Return ownership of system trait object (if entity still exists)
         if let Ok(mut entity) = self.get_entity_mut(id.entity) {
             entity.insert::<RegisteredSystem<I, O>>(RegisteredSystem {
                 initialized,
                 system,
             });
         }
+
+        // Run any commands enqueued by the system
+        self.flush();
```

Comment:
I thought commands were only run at sync points? Not after any system has been run?

Reply:
All good I mostly asked for my own understanding.